docs/_posts/2024-06-27-dynamic-partitioning.md

```diff
@@ -0,0 +1,74 @@
+---
+layout: post
+title: "Dynamic Partitioning in Modin"
+categories: misc
+permalink: /gh_page_3.html
+author: Retribution98
+---
+
+#### Annotation
+
+In 0.30.0 version Modin introduced dynamic partitioning for DataFrames
+having excessive amount of partitions into the `Map` and `TreeReduce` operators.
+This allows to boost performance for certain operations by combining some partitions
+into virtual ones and utilizing CPU cores more efficiently.
+
+### Dynamic Partitioning
+
+As you know, Modin splits a DataFrame into partitions to distribute computation.
+By default, Modin uses the number or CPUs available in the system for this purpose.
+When an operation is executed, the required partitions are sent to remote tasks and processed in parallel.
+To achieve optimal performance it's essential to fully utilize all available CPUs.
+Therefore, Modin splits the DataFrame into partitions along both axes,
+sometimes resulting in an `N × N` partition grid (where N represents the number of CPUs).
+
+This approach works well when the number of partitions close to the number of CPUs.
+However, there are the cases when this method is not that effective. For straightforward operations like `map`
+on a DataFrame that has the number of partitions greatly exceeding the number of CPUs,
+using this partitioning strategy can lead to performance issues due to an excessively large number of remote tasks.
+
+To address this problem, a new approach called `Dynamic Partitioning` has been implemented in Modin 0.30.0.
+The main idea behind `Dynamic Partitioning` is to combine some partitions into virtual ones, thereby reducing the overall number of remote tasks.
+
+### Using Dynamic Partitioning in Modin
+
+You don't need to set any configuration variables to enable this approach.
+Simply update to Modin 0.30.0 or higher.
+
+### Performance result
+
+| Operation | Modin 0.29.0 | Modin 0.30.0 | Ratio |
+| --------- | ------------ | ------------ | ----- |     
+| abs       | 5.768716335  | 1.559553780  | 3.70  |
+| map       | 5.665995907  | 1.663878210  | 3.41  |
+| isna      | 4.371069111  | 1.041565318  | 4.20  |
+| notna     | 4.149922594  | 1.276469827  | 3.25  |
+| round     | 4.789841156  | 1.581816196  | 3.03  |
+| replace   | 4.871268023  | 1.442099884  | 3.38  |
+| count     | 5.163318828  | 1.835885521  | 2.81  |
+| sum       | 5.351826966  | 1.907279816  | 2.81  |
+| prod      | 5.186810397  | 2.101620920  | 2.47  |
+| any       | 5.251107819  | 1.860132668  | 2.82  |
+| all       | 5.724503774  | 1.716603592  | 3.33  |
+| max       | 5.307218991  | 1.764660481  | 3.01  |
+| min       | 5.537900437  | 1.803861558  | 3.07  |
+| mean      | 6.400612667  | 2.005258847  | 3.19  |
+
+To test the performance boost of Dynamic Partitioning, we generated a wide DataFrame with a shape of (20000, 5000).
+As you can see, the new Modin version provides significant speed-up for DataFrames with a large number of columns.
```

Comment:
```suggestion
To test the performance boost of Dynamic Partitioning we generated a wide DataFrame with a shape of (20000, 5000), which result in a large amount of partitions.
As you can see, Modin 0.30.0 provides decent speed-up by benefiting from `Dynamic Partitioning`.
```

Reply:
>To test the performance boost of Dynamic Partitioning we generated a wide DataFrame with a shape of (20000, 5000), which result in a large amount of partitions.
I am not entirely sure if this is absolutely correct, as the number of partitions depends not only on the size of the data. Additionally, I would like to retain the mention of a ‘large number of columns,’ as it provides an easier way to understand the scenarios where this approach works.

Reply:
Even when the number of columns is large, the user may set `MinColumnPartitionsSize` config to a respective value so the number of partitions can get low along the columns. I think we should explicitly mention "large amount of partitions" as I put in the suggestion.

Reply:
Yes, it is possible but I am not sure that `MinColumnPartitionsSize` is a common parameter. 
As you said before, we don't need to add low level details here, but how can user work with the wide Dataframe? In my opinion, using DataFrame with large count of columns is the main way. I think we can add information, that  `MinColumnPartitionsSize` can impact on performance and add link to it in Modin docs.

Reply:
The main point here is "large number of partitions" but not "large number of columns" because even if we have a low number of columns, we can still have a large number of partitions along rows.

Reply:
If a DataFrame has a low number of columns, it usually give not so many partitions, because by the default it no more than CPU count row partitions. (Nx1, or Nx2). The main idea of the resolved issue is working with a large count of columns(NxN).  Of course, it depends on `MinColumnPartitionsSize` value, but it is not most important thing.

I understand, that count of partitions is very important, so it thoroughly explained in the previous paragraphs. But I want keep here a easy way for beginner user to understand which Dataframe is wide.

Reply:
Even if a Dataframe has a low number of columns, there are certain operations in Modin that can lead to an excessive amount of partitions both along rows and columns. In the suggestion above we are mentioning both "wide dataframe" and "large amount of partitions" so I think we give users the understanding what dataframe is wide (providing the shape) and what dataframe can have excessive partitions.

Reply:
Ok, you persuaded me. Thanks for your explanation